src/storage/lock_manager/lock_wait_context.rs

```diff
@@ -128,37 +124,48 @@ impl<L: LockManager> LockWaitContext<L> {
     pub fn get_callback_for_blocked_key(&self) -> PessimisticLockKeyCallback {
         let ctx = self.clone();
         Box::new(move |res| {
-            ctx.finish_request(res);
+            ctx.finish_request(res, false);
         })
     }
 
     /// Get the callback that's used to cancel a lock-waiting request. Usually
     /// called by
     /// [`WaiterManager`](crate::server::lock_manager::WaiterManager) due to
     /// timeout.
+    ///
+    /// This function is assumed to be called when the lock-waiting request is
+    /// queueing but canceled outside, so it includes an operation to actively
+    /// remove the entry from the lock waiting queue.
     pub fn get_callback_for_cancellation(&self) -> impl FnOnce(StorageError) {
         let ctx = self.clone();
         move |e| {
-            ctx.finish_request(Err(e.into()));
+            ctx.finish_request(Err(e.into()), true);
         }
     }
 
-    fn finish_request(&self, result: Result<PessimisticLockRes, SharedError>) {
-        let ctx_inner = if let Some(inner) = self.shared_states.ctx_inner.lock().take() {
-            inner
+    fn finish_request(&self, result: Result<PessimisticLockRes, SharedError>, is_canceling: bool) {
+        if is_canceling {
+            let entry = self
+                .lock_wait_queues
+                .remove_by_token(&self.shared_states.key, self.shared_states.lock_wait_token);
+            if entry.is_none() {
+                // Already popped out from the queue so that it will be woken up normally. Do
+                // nothing.
+                return;
+            }
         } else {
-            debug!("double invoking of finish_request of LockWaitContext";
-                "start_ts" => self.start_ts,
-                "for_update_ts" => self.for_update_ts
-            );
-            return;
-        };
+            // TODO: Uncomment this after the corresponding change of
+            // `LockManager` is done. self.lock_wait_queues.
+            // get_lock_mgr()     .remove_lock_wait(ctx_inner.
+            // lock_wait_token);
```

Comment:
Will we remove the entry from the queue in this branch? In other words, I'm wondering whether the `finished` field in `LockWaitContextSharedState` is still necessary since there will be no stale entried in the queue.

Reply:
This branch is only used when the entry is popped out from the queue and finished. Therefore the entry doesn't exist in the queue but may exist in the `WaiterManager`.
The `finished` field is not necessary now and can be removed. I'll remove it.